### ISSUE TITLE
better support for diploid genotypes

### ISSUE THREAD

**Issue by mah11** [pombase/canto#976]
I hope this isn't too scary.

I have come to the disquieting realisation that we ought to do something a bit sophisticated to capture diploid genotypes. Previously Antonia stuck them in as "other", and more recently we talked about just putting in alleles and some text in the "background" box. But the latter will fall afoul of Canto's clever habit of not adding duplicate alleles in a genotype (which is a good thing)

Our latest wheeze:

Add a tickbox for "diploid" (unticked by default) somewhere convenient on the "add a genotype" page.

If ticked,
- add "diploid" to the genotype description by magic;
- make sure canto collects details for two copies of every gene in the genotype;
- allow the genotype to include two identical alleles for any gene (this will be the case for a diploid that is homozygous for the gene).

Antonia and I tossed around a few ideas for how the alleles would work. Could we do something with the allele details popup like this?
- Enter all the usual stuff for one allele, then see two buttons to proceed (plus the usual 'cancel'):
  - Homozygous (use these details for both copies of yfg1)
  - Heterozygous (add this allele and then adde the other)
- If 'homozygous' is ticked, add two copies of the allele and close the popup. If 'heterozygous' is ticked, return to the allele details popup to gather info on the second allele.

I'll work on the button wordings if this is feasible. Just how pie-in-the-sky is it?

**Comment by kimrutherford** [comment]
> But the latter will fall afoul of Canto's clever habit of not adding duplicate alleles in a genotype

It does allow that at the moment, which is probably not good.  See the screenshot below.

> Just how pie-in-the-sky is it?

I think it would work in Canto and Chado.  I'd like to have a chat about it before promising anything though.

![screenshot 2016-05-25 15 00 47](https://cloud.githubusercontent.com/assets/90474/15526932/9a0c55e8-2289-11e6-902e-f2827586bb70.png)

**Comment by ValWood** [comment]
We should also discuss the formalization for diploids (how they should be displayed on the gene pages
simple wee1+/wee1-
but could get  complex with multi gene diploids.

Q What do others do? and standards?

**Comment by mah11** [comment]
This is a long-solved problem - just look at the genotypes in any paper that deals with diploids. It's even generally consistent between organisms. List each pair of alleles, with spaces between genes.

e.g. wee1+/wee1-50 rad2Δ/rad2Δ cdc2-33/cdc2-33

It's also quite tangential to the actual point of this ticket, which is what Canto allows for genotypes involving wild type (sequence) alleles. This ticket is not about how anything is displayed.

**Comment by ValWood** [comment]
yeah its not complicated at all....

**Comment by ValWood** [comment]
we should think about this with the genotype interface work.

**Comment by kimrutherford** [comment]
We should talk about this again as soon as everyone has time.  It's the next big task for FlyBase-Canto.

**Comment by vmt25** [comment]
Here goes a table with examples of diploid genotypes frequently used by FlyBase - https://docs.google.com/spreadsheets/d/16fjZJzYXsrxgr-Oz6iMgtyz_S3PnG89UHcR4gU-rZ5Q/edit#gid=1866375701

Two alleles affecting the same gene in homologous chromosomes (locus) are separated by "/"; alleles separated by "," are at different chromosomes or at different chromosome locations.

**Comment by ValWood** [comment]
Our first mock-up (from the haploid by default perspective)

![mockup](https://user-images.githubusercontent.com/7359272/57073928-6aae7800-6cda-11e9-846c-09de6148bd76.jpg)


![mockup](https://user-images.githubusercontent.com/7359272/57073909-5cf8f280-6cda-11e9-8d3a-bef869c2b980.jpg)

**Comment by ValWood** [comment]
It's a bit rough  but a start
- We would allow configuration of a default haploid or diploid designation per species 
 for mixed species anotation
- we haven't discussed multiploidy, we will probably need to make some assumptions/generalizations  in these cases for plants
- we envisage that FlyBase would preload into diploid tables in "diploid by default " mode
- we also discussed edge cases like a knock-in extra copy for a diploid but this might be rare enough to be worth doing directly in edit interface after diploid created
-  in 'haploid' mode the 'create diploid' button would only be available when a single allele selected (so that all other genomic copies added in diploid interface are allelic) therefore
note that absent should be x-d out in the figure above 

did we discuss anything else? I didn't make notes since I was doing the mock-up

@Antonialock  does this make sense to you?

**Comment by kimrutherford** [comment]
To help with discussion, I've added the code to make a dialog.  The "Create diploid" button opens the dialog but the OK button in the dialog doesn't do anything.  :-)

I've updated the flybase-test Canto with the changes:
https://curation.pombase.org/flybase-test/curs/d784e38657302d54/genotype_manage

![combine-genotypes-button-2](https://user-images.githubusercontent.com/90474/57209299-1da90980-702c-11e9-8ea7-03c8427685fd.png)
![diploids-1](https://user-images.githubusercontent.com/90474/57209301-20a3fa00-702c-11e9-986f-ea82becf87a9.png)

**Comment by kimrutherford** [comment]
I've added "homozygous" to the dialog.  We should chat about how the homozygous and wild type options should be displayed.  "vn[C221] / homozygous" looks a bit odd.

![diploids-2](https://user-images.githubusercontent.com/90474/57282338-f1ab8800-7100-11e9-9b56-2105ecd22118.png)

The "OK" button now stores the selected alleles as a pair in the database.  They are shown like this at the moment:

![diploids-4](https://user-images.githubusercontent.com/90474/57282649-9af27e00-7101-11e9-883b-7d0a93df0319.png)

They can't be included in multi-allele genotypes yet.

**Comment by kimrutherford** [comment]
For pombe: how does expression work with diploid loci?  Currently we have the expression in a separate column.  Will that work for diploids?  Or do we need to remove the expression column and display something like this?:
```
  cdc2-1w (G146D aa) Overexpressed / cdc2+ Knockdown
```

**Comment by kimrutherford** [comment]
For FlyBase: what is your usual way of displaying homozygous loci?
Canto currently displays "vn[10567] / vn[10567]" which doesn't seem great.

And how should it look when one allele is the wild type allele?
Current it's displayed as: "vn[C221] / vn+"

I've updated the flybase-test Canto if you'd like to play around:
  https://curation.pombase.org/flybase-test/curs/527034fd25f7711e/genotype_manage

**Comment by kimrutherford** [comment]
> They can't be included in multi-allele genotypes yet.

That sort-of works now as long as all the loci are diploid:

![diploids-4 5](https://user-images.githubusercontent.com/90474/57287809-6a641180-710c-11e9-8345-6cbc8a03acf0.png)

![diploids-5](https://user-images.githubusercontent.com/90474/57287820-6cc66b80-710c-11e9-8e44-14b1190f9bd4.png)

**Comment by kimrutherford** [comment]
> That sort-of works now as long as all the loci are diploid

That's fixed I think:

![diploids-8](https://user-images.githubusercontent.com/90474/57292870-275b6b80-7117-11e9-9a7f-ca503a45b801.png)

But the display of genotype in the annotation tables needs fixing:

![diploids-7](https://user-images.githubusercontent.com/90474/57292784-ebc0a180-7116-11e9-8eea-4ee148054337.png)

The genotype edit page doesn't understand diploids either:

![diploids-9](https://user-images.githubusercontent.com/90474/57293108-d5ffac00-7117-11e9-904b-cd89cfd05dc5.png)

**Comment by mah11** [comment]
Cooooool to see this happening at last. I'll try to find time for testing ...

Comments in the meantime:

> We should chat about how the homozygous and wild type options should be displayed. "vn[C221] / homozygous" looks a bit odd.

Yes, and the reason it's odd is that "/" separates two allele names, but "homozygous" isn't an allele name or description. Possible solutions:

- put the allele name (and/or description) in the pulldown instead of "homozygous", or

- put the "/" in the pulldown for the options that are allele names or descriptions, but not for the "homozygous" option.

Or maybe someone has another idea ...

Another thing: unless Drosophila nomenclature does something bizarrely non-standard, "vn+" and "wild type" refer to the same thing. Are they both included on purpose?

> For pombe: how does expression work with diploid loci? Currently we have the expression in a separate column. Will that work for diploids? Or do we need to remove the expression column and display something like this?:
> 
>    cdc2-1w (G146D aa) Overexpressed / cdc2+ Knockdown

I think we'll need this new display. (Sorry for the extra work.)


> For FlyBase: what is your usual way of displaying homozygous loci?
> Canto currently displays "vn[10567] / vn[10567]" which doesn't seem great.
> 
> And how should it look when one allele is the wild type allele?
> Current it's displayed as: "vn[C221] / vn+"

I don't presume to speak for FlyBase, but these actually look correct, and therefore fine, from my "person with a clue about general genetics but not a Drosophila expert" point of view.

**Comment by kimrutherford** [comment]
> Another thing: unless Drosophila nomenclature does something bizarrely non-standard, "vn+" and
> "wild type" refer to the same thing. Are they both included on purpose?

I'll fix that.  It's happening because there was already a wild type single allele genotype in the session.  The code always puts "wild type" in the dropdown then it adds all the alleles from the session. So in that case we got two.

**Comment by kimrutherford** [comment]
> We should chat about how the homozygous and wild type options should be displayed.

I've had another go at this.  I think this is clearer:
![diploids-10](https://user-images.githubusercontent.com/90474/57339909-a9837880-7187-11e9-9944-eaf9a50c2d23.png)

It also looks better when there are no other alleles:
![diploids-12](https://user-images.githubusercontent.com/90474/57339922-bacc8500-7187-11e9-92bc-dc76019331ab.png)

**Comment by mah11** [comment]
I like it!

**Comment by vmt25** [comment]
The version in the test server looks great.

> For FlyBase: what is your usual way of displaying homozygous loci?
> Canto currently displays "vn[10567] / vn[10567]" which doesn't seem great.
> 
> And how should it look when one allele is the wild type allele?
> Current it's displayed as: "vn[C221] / vn+"

> I don't presume to speak for FlyBase, but these actually look correct, and therefore fine, from my "person with a clue about general genetics but not a Drosophila expert" point of view.

Both representations (and Midori's assessment) are correct.
"vn[10567] / vn[10567]" is the usual way of displaying homozygosity.
"vn[C221] / vn+" would be fine but "vn[C221] / vn[+]" would be better, especially regarding our future changes to genotype storage in the database.

> I've had another go at this. I think this is clearer:
> 
> It also looks better when there are no other alleles:

I agree that the current menu looks better than the first instance - please keep it.


Transheterozygous ("vn[10567] / vn[C221]"), and arguably heterozygous ("vn[C221] / vn+" ), genotypes technically are not single allele genotypes, although it may be useful to keep them in the same table. We could discuss this next week.

**Comment by mah11** [comment]
> "vn[C221] / vn[+]" would be better

I disagree. The wild type allele designation is "vn+", not "vn[+]", and it should be displayed correctly.

I'm assuming that things like "vn[C221]" and "vn[10567]" are fly-specific designations, and for pombe (and other species) we would use whatever the allele names are anyway, e.g. "wee1-50 / wee1-50" or "wee1-50 / wee1+".

>  ("vn[10567] / vn[C221]"), and arguably heterozygous ("vn[C221] / vn+" ), genotypes technically are not single allele genotypes

No argument - there are indisputably two alleles in each of these. We did briefly raise the possibility of using a phrase like "single gene of interest" for the table in question, but I don't think we got as far as non-clunky wording.

**Comment by kimrutherford** [comment]
> Another thing: unless Drosophila nomenclature does something bizarrely non-standard, "vn+" and
> "wild type" refer to the same thing. Are they both included on purpose?

That's fixed now: the wild type allele it appears only once in the dialog even if you have an existing WT allele in the session.

![diploids-13](https://user-images.githubusercontent.com/90474/57432035-c653a500-7288-11e9-9d17-16242ba7ee0c.png)

**Comment by vmt25** [comment]
> > "vn[C221] / vn[+]" would be better
> 
> I disagree. The wild type allele designation is "vn+", not "vn[+]", and it should be displayed correctly.
> 
> I'm assuming that things like "vn[C221]" and "vn[10567]" are fly-specific designations, and for pombe (and other species) we would use whatever the allele names are anyway, e.g. "wee1-50 / wee1-50" or "wee1-50 / wee1+".

Superscripts/square brackets are the formal way FlyBase, and the fly community, designates mutant alleles.

For mutant alleles, FlyBase would appreciate keeping the brackets, as it immediatly guides to the allele name and helps when there are paralogues or genes with similar symbols (and symbols with numbers).

For wild-type "alleles", puting the + within brackets would be helpful for FlyBase's purposes, but I'm sure we can work around it.

**Comment by mah11** [comment]
> For wild-type "alleles", puting the + within brackets would be helpful for FlyBase's purposes

OK, so yfg[+] for flies, and yfg1+ for pombe, and YFG1+ for budding yeast ... maybe this should be configurable in Canto (sorry Kim!)

**Comment by kimrutherford** [comment]
> maybe this should be configurable in Canto

No problem.

**Comment by kimrutherford** [comment]
I've fixed the bug that was disabling the OK button when only the homozygous and wild type options were available.

**Comment by kimrutherford** [comment]
> OK, so yfg[+] for flies, and yfg1+ for pombe, and YFG1+ for budding yeast ... maybe this should be
> configurable in Canto (sorry Kim!)

Turns out that the wild type allele format is already configurable, I just need to document how it's done. I've made a separate issue for that: #1876

Still to implement:

- fix display of diploids in genotype/phenotype annotation tables
- fix display on genotype edit page
- remove expression column and put it beside the genotype name, see: https://github.com/pombase/canto/issues/976#issuecomment-489976815

**Comment by kimrutherford** [comment]
> Turns out that the wild type allele format is already configurable,

It wasn't working for the wildtype alleles create by the diploid dialog.  I've fix that now.

**Comment by kimrutherford** [comment]
> > They can't be included in multi-allele genotypes yet.
>
> That sort-of works now as long as all the loci are diploid:

I did some further testing and found that it goes quite badly wrong if two diploids in the mutli-allele genotype have an allele in common.  It's not the commonest case but when I looked at the code I realised the current implementation is likely to cause other bugs.

I'll work on fixing that tomorrow.  Then I'll get back to the "Still to implement" list:
- fix display of diploids in genotype/phenotype annotation tables
- fix display on genotype edit page
- remove expression column and put it beside the genotype name, see: https://github.com/pombase/canto/issues/976#issuecomment-489976815

**Comment by kimrutherford** [comment]
> I did some further testing and found that it goes quite badly wrong if two diploids in the mutli-allele
> genotype have an allele in common.

That's fixed.  The code is simpler now so I'm more confident that it's correct.

Sorry for the huge list of Git commits attached to this issue.  I looks like I've done lots of work but they are mostly duplicates.  I got myself in a muddle involving Git branches.

**Comment by kimrutherford** [comment]
> fix display on genotype edit page

It's a bit better now.  The alleles for each diploid are grouped together.  The "Add allele ...", "Edit" and "Remove" buttons work except that "Add allele ..." adds the new allele as a haploid.  We should talk about what to do about that, if anything.

Submitting the edits with "Finish" still messes up the genotype.  I'll work on that next.

(We probably could do with improvement the title/header of this page to make it clear that it's editing mode)

![diploid-edit-1](https://user-images.githubusercontent.com/90474/58451896-740bf280-8169-11e9-905a-5230a9104fae.png)

**Comment by ValWood** [comment]
Is this in the pombase test server, if so I can take it for a spin.

I'm trying now but test is really, really slow? Like a cow....

**Comment by kimrutherford** [comment]
I can't see any problems with the server and the test Canto is fast for me.  Could it be your network?

**Comment by ValWood** [comment]
I see it isn't in test yet.
It looks perfect from the above. if you add to test we can double check with some of our real examples.

**Comment by ValWood** [comment]
> Could it be your network?

quite possibly. Nothing else seems so slow though...

**Comment by kimrutherford** [comment]
> if you add to test we can double check with some of our real examples.

I'm updating test to add diploid support.  I'll let you know when that's done.

**Comment by kimrutherford** [comment]
> I'm updating test to add diploid support. I'll let you know when that's done.

It's done now.  You'll need to be logged in as admin to see the "Create diploid..." button on the genotype management page.

**Comment by kimrutherford** [comment]
> Submitting the edits with "Finish" still messes up the genotype. I'll work on that next.

That's working much better now, but needs testing.

Diploid mode is enabled in the flybase-test Canto and in the PomBase test Canto (https://curation.pombase.org/test/).

The PomBase test Canto is temporarily offline for a database upgrade.  It should be back in 10-ish minutes.

**Comment by ValWood** [comment]
I could not see the "diploid" option when I went to test...

**Comment by kimrutherford** [comment]
> I could not see the "diploid" option when I went to test...

Were you logged in?

**Comment by ValWood** [comment]
yep, checked again, still don't see it. even when I select a single allele mutant (not sure if this is necessary?)

**Comment by kimrutherford** [comment]
Sorry, I was thinking of the synonyms - you don't need to be logged into to see the "create diploid" button in the test Canto.  It's OK for me.  For example: https://curation.pombase.org/test/curs/35bfa0019d25c550/genotype_manage

**Comment by ValWood** [comment]
OK got it. For some reason I thought it was in the live version. i will test in test...

**Comment by ValWood** [comment]
It seems to work well.

a few of small things:
- [ ]  It's a little confusing if you don't know what's happening that you cant combine multiple single alleles into a diploid. therefore, I think the pop-up should say "create diploid allele" rather than "create diploid"

- [ ] it feels a bit weird that the create diploid button does not remain under the single locus 'haploid alleles" but appears under the diploid table, once you have created one diploid.

- [ ]  We should probably prevent users from  selecting.creating multi allele genotypes from a mixture of "haploid" and "diploid" alleles. Actually, this is biologically possible but would be unlikely for fission yeast. Maybe a warning  "you are about to .....are you sure".  We could discuss this one....

**Comment by kimrutherford** [comment]
> I think the pop-up should say "create diploid allele" 

Maybe "Create diploid locus"?

> it feels a bit weird that the create diploid button does not remain under the single locus 'haploid alleles"
> but appears under the diploid table, once you have created one diploid.

OK, I'll work on that.

> We should probably prevent users from selecting.creating multi allele genotypes from a mixture of
> "haploid" and "diploid" alleles. Actually, this is biologically possible but would be unlikely for fission
> yeast. Maybe a warning "you are about to .....are you sure". We could discuss this one....

I think for fly at least it would have to be a warning only.

**Comment by ValWood** [comment]
> Maybe "Create diploid locus"?
doh yes. 
That would be much clearer

**Comment by ValWood** [comment]
> I think for fly at least it would have to be a warning only.

Yes you are right, and also for lots of pathogens. Ignore this for now. If the button moves and it is clear that you are creating a diploid locus, this probably  won't be an issue

**Comment by kimrutherford** [comment]
> a few of small things:

I've changed the dialog title and the "Create diploid" button now stays under the single allele genotypes table.

**Comment by Antonialock** [comment]
I don't like that the "action: create diploid" and "Actions:  combine selected genotypes" are so far away from each other...it's also not obvious what the difference is.

**Comment by ValWood** [comment]
One is to create diploids and one to combine genotypes?


Would it be better if the combine genotypes action was repeated under both tables?

**Comment by Antonialock** [comment]
I just wonder if the wording is intuitive unless you know whats going on...

**Comment by ValWood** [comment]
which wording? "create diploid locus"?

We can change it if you have suggestions?

**Comment by kimrutherford** [comment]
> the "Still to implement" list:
> -   fix display of diploids in genotype/phenotype annotation tables
> -   fix display on genotype edit page
> -   remove expression column and put it beside the genotype name

This list is all done now.

I think we could almost put diploids into production.  But lot of code has been modified to implement this so I think it's worth doing a little bit more testing. If you could make a few genotypes with diploids in the test Canto and add a few annotations that would be very helpful.  If you could do that and let me know the session ID, I'll export the annotation and test loading into Chado.

If you find bugs please open new issues.

The last thing to decide in this issue is where to put the "Create diploid" and Combine buttons.

**Comment by ValWood** [comment]
Can everybody test and suggest any refinements.
https://github.com/pombase/curation/issues/2418
VAl

**Comment by kimrutherford** [comment]
One change I'm thinking of is to change the "Diploid 1" and "Diploid 2" etc. labels on the edit page to say "Locus 1" and "Locus 2".  Does that make sense?

Also would "Create diploid locus" be a better label for the "Create diploid" button?

**Comment by ValWood** [comment]
> Also would "Create diploid locus" be a better label for the "Create diploid" button?
I think so....

**Comment by kimrutherford** [comment]
>>Also would "Create diploid locus" be a better label for the "Create diploid" button?
>
> I think so....

OK, I've done that.

**Comment by vmt25** [comment]
Sorry for the late reply

'Create diploid locus' reads better.

Cannot see the locus 1, locus 2, .... Don't think it is necessary but also does not harm.

**Comment by vmt25** [comment]
'Allele' and 'annotations' column labels on genotype tables are swaped.

![Screenshot 2019-06-11 at 11 03 00](https://user-images.githubusercontent.com/33721108/59263502-14b5e400-8c39-11e9-8757-497225d68350.png)

**Comment by kimrutherford** [comment]
> 'Allele' and 'annotations' column labels on genotype tables are swaped.

Fixed!  Sorry about that.  I forgot to test a previous change in FlyBase mode.

**Comment by kimrutherford** [comment]
> Cannot see the locus 1, locus 2, .... Don't think it is necessary but also does not harm.

I haven't made that change yet.  I'm waiting for feedback.

**Comment by vmt25** [comment]
The name of heterozygous genotypes (mutant/wild-type) shows in a weird format:

![Screenshot 2019-06-11 at 11 58 09](https://user-images.githubusercontent.com/33721108/59266722-38305d00-8c40-11e9-9426-cca497486b97.png)

Using the example above, I thought it was going to be something like Alk[1]/Alk[+] or Alk[1]/Alk+ or simply Alk[1]/+

**Comment by kimrutherford** [comment]
> The name of heterozygous genotypes (mutant/wild-type) shows in a weird format:

That's fixed.  It will now show something like: `Psc[e24] / Psc[+]`

**Comment by vmt25** [comment]
Hopefully a small request:
set multi-allele genotype to always present accessory alleles at the bottom

E.g., in the example below, make Scer\GAL4[Mef2.PU](accessory) the last of the three alleles.

![Screenshot 2019-06-11 at 14 03 00](https://user-images.githubusercontent.com/33721108/59274149-a7fb1380-8c51-11e9-8089-a589e97208f0.png)

**Comment by vmt25** [comment]
> That's fixed. It will now show something like: `Psc[e24] / Psc[+]`

Thanks

**Comment by ValWood** [comment]
I'm closing this ticket. New small tickets for outstanding tasks. If you have something in this ticket that isn't done, please open a new one.
